File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        //brute force will be O(n) both time and space
        // int n = nums.size();
        // unordered_map<int,int> mapi;

        // for(int i = 0; i < n; i++)mapi[nums[i]]++;
        // int ans = 0;
        // for(auto p : mapi){
        //     if(p.second > n/2){
        //         ans = p.first;
        //         break;
        //     }
        // }
        // return ans;

        // Better in constant space
        // This Moore's Voting Algorithm
        // If the majority element is there then it must be in majority in some portion of the array
        int n = nums.size();
        int count = 0;
        int element;

        for(int i = 0; i < n; i++){
            if(count == 0){
                element = nums[i];
                count = 1;
            }
            else if(nums[i] == element)count++;
            else count--;

        }
        return element;
    }
};

File: 0169-majority-element/0169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        // 1. consider your majority element as +ve and other as -ve
        // 2. If majority exist then +ve will more than the -ve
        // 3. The +ve can be together or in some portion throughout the array like +++--++ 
        // 4. At the end your ele variable end up with taking majority element with count > 0
        // 5. If we are not guaranteed that majority element exist or not so it doesn't mean that count > 0 so majority will exist then in that case you have to frq of ele variable then check > [n/2]
        int count = 1;
        int ele = nums[0];

        for(int i = 1; i < nums.length; i++){
            if(count == 0){
                ele = nums[i];
                count = 1;
            }else if(ele == nums[i])count++;
            else count--;
        }

        return ele;
    }
}